File: app/src/student/java/com.apptronix.nitkonschedule.student/ui/CoursesFragment.java
package com.apptronix.nitkonschedule.student.ui;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import androidx.appcompat.app.AlertDialog;
import android.text.InputType;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.Toast;

import com.apptronix.nitkonschedule.R;
import com.apptronix.nitkonschedule.model.Assignment;
import com.apptronix.nitkonschedule.model.AssignmentResponse;
import com.apptronix.nitkonschedule.model.Course;
import com.apptronix.nitkonschedule.model.User;
import com.apptronix.nitkonschedule.rest.ApiClient;
import com.apptronix.nitkonschedule.student.adapter.CoursesAdapter;
import com.apptronix.nitkonschedule.student.adapter.TestsAdapter;
import com.apptronix.nitkonschedule.student.data.DBContract;
import com.apptronix.nitkonschedule.student.rest.ApiInterface;
import com.apptronix.nitkonschedule.student.service.UploadService;

import java.io.IOException;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

/**
 * Created by Maha Perriyava on 4/6/2018.
 */

public class CoursesFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor>{

    private CoursesFragment.OnFragmentInteractionListener mListener;
    CoursesAdapter coursesAdapter;
    User user;
    EditText input;

    public CoursesFragment() {
        // Required empty public constructor
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {

        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_courses, container, false);
        ListView courseList = (ListView)rootView.findViewById(R.id.coursesListView);
        coursesAdapter = new CoursesAdapter(getActivity(),null);
        courseList.setAdapter(coursesAdapter);
        courseList.setEmptyView(rootView.findViewById(R.id.empty));
        getLoaderManager().initLoader(0,null,this);
        FloatingActionButton fab = rootView.findViewById(R.id.enrollCourse);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AlertDialog.Builder alertDialog = new AlertDialog.Builder(getActivity());
                alertDialog.setTitle("Enroll Course");

                // Setting Dialog Message
                alertDialog.setMessage("Enter Course code");
                input = new EditText(getActivity());
                input.setGravity(Gravity.CENTER);
                input.setInputType(InputType.TYPE_TEXT_FLAG_CAP_CHARACTERS);
                LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.MATCH_PARENT);
                input.setLayoutParams(lp);
                alertDialog.setView(input);
                alertDialog.setIcon(R.drawable.ic_school_black_24dp);

                alertDialog.setPositiveButton("ENROLL",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog,int which) {
                                // Write your code here to execute after dialog

                                Bundle bundle = new Bundle();
                                bundle.putString("content","enrollCourse");
                                bundle.putSerializable("parcel", new Course(input.getText().toString()));
                                Intent intent = new Intent(getContext(), UploadService.class);
                                intent.putExtra("bundle",bundle);
                                Timber.i("starting service");
                                getActivity().startService(intent);

                            }
                        });
                alertDialog.setNegativeButton("CANCEL",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                // Write your code here to execute after dialog
                                dialog.cancel();
                            }
                        });

                // closed

                // Showing Alert Message
                alertDialog.show();

            }
        });
        return rootView;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof TestsFragment.OnFragmentInteractionListener) {
            mListener = (CoursesFragment.OnFragmentInteractionListener) context;
            mListener.onFragmentInteraction("Courses ");
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public Loader onCreateLoader(int id, Bundle args) {

        return new CursorLoader(getActivity(),
                DBContract.CourseEntry.CONTENT_URI,
                null,
                null,
                null,
                null);
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        coursesAdapter.swapCursor(data);
    }


    @Override
    public void onLoaderReset(Loader loader) {
        coursesAdapter.swapCursor(null);
    }

    public interface OnFragmentInteractionListener {

        void onFragmentInteraction(String token);
    }
}

File: app/src/teacher/java/com/apptronix/nitkonschedule/teacher/ui/fragments/CoursesFragment.kt
package com.apptronix.nitkonschedule.teacher.ui.fragments

import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.appcompat.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import com.apptronix.nitkonschedule.R
import com.apptronix.nitkonschedule.teacher.adapter.CoursesAdapter
import com.apptronix.nitkonschedule.teacher.data.DBContract
import com.apptronix.nitkonschedule.teacher.ui.NewCourseActivity
import com.apptronix.nitkonschedule.teacher.ui.ResourcesActivity
import kotlinx.android.synthetic.teacher.fragment_courses.view.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import timber.log.Timber


class CoursesFragment : Fragment() , LoaderManager.LoaderCallbacks<Cursor> {

    override fun onLoaderReset(loader: Loader<Cursor>) {
        coursesAdapter.swapCursor(null)
    }

    private var mListener: OnFragmentInteractionListener? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments !=
                null) {

        }
    }

    lateinit var  cursor: Cursor
    lateinit var coursesAdapter: CoursesAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment\

        val rootView = inflater!!.inflate(R.layout.fragment_courses, container, false)
        val coursesList = rootView.findViewById<View>(R.id.coursesListView) as ListView

        coursesAdapter= CoursesAdapter(activity,null)
        coursesList.adapter = coursesAdapter
        loaderManager.initLoader(0, null, this)

        coursesList.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            cursor!!.moveToPosition(position)
            val builder = AlertDialog.Builder(activity!!)
            builder.setTitle(R.string.pick_action)
                    .setItems(R.array.course_actions_array) { dialog, which ->
                        Timber.i("touch pos %d", which)
                        when (which) {
                            0 -> {//resources
                                val intent = Intent(activity, ResourcesActivity::class.java)
                                intent.putExtra("course", cursor.getString(cursor.getColumnIndex(DBContract.CourseEntry.COLUMN_COURSE)))
                                startActivity(intent)
                            }
                            1 -> {//tests

                            }
                            2 -> {//assignments
                            }
                        }
                    }
            val dialog = builder.create()
            dialog.show()
        }
        rootView.addCourse.setOnClickListener {
            val intent = Intent(activity, NewCourseActivity::class.java)
            startActivity(intent)
        }

        return rootView
    }


    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
            mListener!!.onFragmentInteraction("Courses")
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {

        return CursorLoader(activity!!,
                DBContract.CourseEntry.CONTENT_URI, null, null, null, null)
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        cursor = data
        coursesAdapter.swapCursor(data)
    }

    // TODO: Update argument type and name
    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(title: String)

        fun onFragmentChange(code: String, title: String)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: MessageEvent) {

        Timber.i(event.message)

        when (event.message) {
            "reload" -> {

                loaderManager.restartLoader(0,null,this)

            }
        }
    }

    class MessageEvent(var message: String)

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }
}